src/main/java/com/jozufozu/flywheel/backend/engine/uniform/LevelUniforms.java

```diff
@@ -0,0 +1,93 @@
+package com.jozufozu.flywheel.backend.engine.uniform;
+
+import org.jetbrains.annotations.Nullable;
+import org.lwjgl.system.MemoryUtil;
+
+import com.jozufozu.flywheel.api.event.RenderContext;
+
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.multiplayer.ClientLevel;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.phys.Vec3;
+import net.minecraft.world.level.Level;
+
+public class LevelUniforms implements UniformProvider {
+	public static final int SIZE = 12 * 4 + 2 * 16;
+
+	@Nullable
+	private RenderContext context;
+
+	@Override
+	public int byteSize() {
+		return SIZE;
+	}
+
+	public void setContext(RenderContext context) {
+		this.context = context;
+	}
+
+	@Override
+	public void write(long ptr) {
+		Minecraft mc = Minecraft.getInstance();
+		ClientLevel level = mc.level;
```

Comment:
This level should be retrieved from the context, not the client singleton.

Reply:
Fixed.